### PR TITLE
better signature and inference for foldLeft on KList

### PR DESCRIPTION
**Pull request by eparejatobes**

### CODE REVIEW

**Comment by eparejatobes** [comment]
This should make possible folding with variable types.

**Comment by eparejatobes** [comment]
@laughedelic please take a look at this I'm a bit tired and I think that what we call foldLeft now is foldRight :laughing:

anyway it works nicely now, you can implement for example concatenation in terms of it.

**Comment by eparejatobes** [comment]
also I think it worthwhile to add a higher priority implicit for the case where you have a "normal" function as argument; in that case you should convert the argument to List/whatever and use collections foldLeft/foldRight.

**Comment by eparejatobes** [comment]
Merging once

- [ ] we have an optimized version for normal `(A,B) => B` functions
- [ ] decide whether this should be foldLeft or foldRight :)
- [ ] review @laughedelic

**Comment by laughedelic** [comment]
ok. I'll review this tomorrow.

**Comment by eparejatobes** [comment]
So what we have now is foldRight. I vote for adding foldLeft and keeping both, without any reverse trick

**Comment by laughedelic** [comment]
I'm doing it

**Comment by laughedelic** [comment]
@eparejatobes take a look, please. I got stuck with this foldLeft. I think it doesn't derive the out type, I guess, it can be solved by some juggling with `FoldLeft.cons`'s type parameters...

**Comment by laughedelic** [comment]
@eparejatobes I think I'm done here. please, review and let's merge it :shipit:

File: src/main/scala/cosas/fns/instances.scala [hunk 1]
```diff
@@ -23,6 +20,18 @@ case object Fn1 {
...
+
+  def f: (In1,In2) => Out
+}
+case class Fn2[A, B, C](val f: (A,B) => C) extends AnyVal with AnyFn2 with DepFn2[A, B, C]
+
+case object Fn2 {
+
+  implicit def appForFn2[A, B, C](df: Fn2[A, B, C]):
+    App2[Fn2[A, B, C], A, B, C] = App2 { (a: A, b: B) => df.f(a, b) }
+}
```
Comment (laughedelic) [inline src/main/scala/cosas/fns/instances.scala]: is this ok? I didn't get what is happening in the case of Fn1 (above), but I think this is enough.

Comment (eparejatobes) [inline src/main/scala/cosas/fns/instances.scala]: I'd rather have the other thing, just to reduce allocations

**Comment by eparejatobes** [comment]
+1 to everything but the fn2 detail

**Comment by eparejatobes** [comment]
Also better to have concat and similar things as compositions if possible.

**Comment by laughedelic** [comment]
> +1 to everything but the fn2 detail

ok. then you change it.

> better to have concat and similar things as compositions if possible

composition of what?

**Comment by eparejatobes** [comment]
LGTM

**Pull request merged**